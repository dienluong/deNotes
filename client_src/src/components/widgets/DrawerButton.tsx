import React from 'react';
import Fab from '@material-ui/core/Fab';
import MenuIcon from '@material-ui/icons/Menu';
import Zoom from '@material-ui/core/Zoom';

function DrawerButton({ visible, clickHandler, className }: { visible: boolean, clickHandler: () => unknown, className: string }) {
  return (
    <div className={ className }>
      <Zoom in={ visible } unmountOnExit>
        <Fab aria-label="notes list drawer" onClick={ clickHandler }>
          <MenuIcon />
        </Fab>
      </Zoom>
    </div>
  );
}

export default DrawerButton;
