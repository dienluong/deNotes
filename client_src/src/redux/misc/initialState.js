import uuid from 'uuid/v1';
import Delta from 'quill-delta';
const _ID_DELIMITER = process.env.REACT_APP_ID_DELIMITER;

// TODO: Is it the best place to define this?
const _rootNode = [
  {
    title: '/',
    subtitle: '',
    uniqid: uuid(),
    get id() {
      return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
    },
    type: 'folder',
    expanded: true,
    children: [],
  },
];

const epoch = 0;
export default {
  notesTree: _rootNode,
  activeNode: {
    id: _rootNode[0].id,
    path: [_rootNode[0].id],
  },
  editorContent: {
    id: '',
    title: '',
    content: '<p><br></p>',
    delta: new Delta(),
    dateCreated: epoch,
    dateModified: epoch,
    readOnly: false,
  },
};
