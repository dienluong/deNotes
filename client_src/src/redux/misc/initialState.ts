import uuid from 'uuid/v4';
import Delta from 'quill-delta';
import { nodeTypes, DEFAULT_ID_DELIMITER} from '../../utils/appCONSTANTS';

const _ID_DELIMITER = process.env.REACT_APP_ID_DELIMITER || DEFAULT_ID_DELIMITER;

const _rootNode: TreeNodeT = {
    title: '/',
    subtitle: '',
    uniqid: uuid(),
    get id() {
      return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
    },
    type: nodeTypes.FOLDER,
    expanded: true,
    children: [],
};

const epoch = 0;
const now = Date.now();
const initialState: AppStateT = {
  userInfo: {
    id: 'default-user',
  },
  notesTree: {
    id: uuid(),
    tree: [_rootNode],
    editMode: false,
    editModeSelectedNodes: [],
    dateCreated: now,
    dateModified: now,
  },
  // activeNode contains the ID of the current active node, and its path, in the tree;
  // the active node can be an item (note) or a folder;
  // if the active node is the root (with no child selected), then ID is constant NONE_SELECTED and path is [NONE_SELECTED]
  activeNode: {
    id: _rootNode.id,
    path: [_rootNode.id],
  },
  editorContent: {
    id: 'default',
    title: '',
    content: '<p><br></p>',
    delta: new Delta(),
    dateCreated: epoch,
    dateModified: epoch,
    readOnly: true,
  },
  modalInfo: {
    type: '',
    props: {},
  }
};

export default initialState;
