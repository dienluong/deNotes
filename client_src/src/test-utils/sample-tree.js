import uuid from 'uuid/v1';
import { nodeTypes } from '../utils/appCONSTANTS';

function randomDateString(start, end) {
  return new Date(start.getTime() +
    Math.random() * (end.getTime() - start.getTime())).toLocaleString();
}

const start = new Date(2018, 5, 1);
const end = new Date();
const _ID_DELIMITER = process.env.REACT_APP_ID_DELIMITER;

export default [
  {
    title: 'react-ui-tree',
    subtitle: randomDateString(start, end),
    uniqid: uuid(),
    get id() {
      return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
    },
    type: nodeTypes.FOLDER,
    expanded: true,
    children: [
      {
        title: 'dist',
        subtitle: randomDateString(start, end),
        uniqid: uuid(),
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
        type: nodeTypes.FOLDER,
        children: [
          {
            title: 'node.js',
            subtitle: randomDateString(start, end),
            uniqid: uuid(),
            get id() {
              return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
            },
            type: nodeTypes.ITEM,
          },
          {
            title: 'react-ui-tree.css',
            subtitle: randomDateString(start, end),
            uniqid: uuid(),
            get id() {
              return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
            },
            type: nodeTypes.ITEM,
          },
          {
            title: 'react-ui-tree.js',
            subtitle: randomDateString(start, end),
            uniqid: uuid(),
            get id() {
              return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
            },
            type: nodeTypes.ITEM,
          },
          {
            title: 'tree.js',
            subtitle: randomDateString(start, end),
            uniqid: uuid(),
            get id() {
              return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
            },
            type: nodeTypes.ITEM,
          },
        ],
      },
      {
        title: 'example',
        subtitle: randomDateString(start, end),
        uniqid: uuid(),
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
        type: nodeTypes.FOLDER,
        children: [
          {
            title: 'app.js',
            subtitle: randomDateString(start, end),
            uniqid: uuid(),
            get id() {
              return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
            },
            type: nodeTypes.ITEM,
          },
          {
            title: 'app.less',
            subtitle: randomDateString(start, end),
            uniqid: uuid(),
            get id() {
              return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
            },
            type: nodeTypes.ITEM,
          },
          {
            title: 'index.html',
            subtitle: randomDateString(start, end),
            uniqid: uuid(),
            get id() {
              return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
            },
            type: nodeTypes.ITEM,
          },
        ],
      },
      {
        title: 'lib',
        subtitle: randomDateString(start, end),
        uniqid: uuid(),
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
        type: nodeTypes.FOLDER,
        children: [
          {
            title: 'node.js',
            subtitle: randomDateString(start, end),
            uniqid: uuid(),
            get id() {
              return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
            },
            type: nodeTypes.ITEM,
          },
          {
            title: 'react-ui-tree.js',
            subtitle: randomDateString(start, end),
            uniqid: uuid(),
            get id() {
              return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
            },
            type: nodeTypes.ITEM,
          },
          {
            title: 'react-ui-tree.less',
            subtitle: randomDateString(start, end),
            uniqid: uuid(),
            get id() {
              return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
            },
            type: nodeTypes.ITEM,
          },
          {
            title: 'tree.js',
            subtitle: randomDateString(start, end),
            uniqid: uuid(),
            get id() {
              return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
            },
            type: nodeTypes.ITEM,
          },
        ],
      },
      {
        title: '.gitiignore',
        subtitle: randomDateString(start, end),
        uniqid: uuid(),
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
        type: nodeTypes.ITEM,
      },
      {
        title: 'index.js',
        subtitle: randomDateString(start, end),
        uniqid: uuid(),
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
        type: nodeTypes.ITEM,
      },
      {
        title: 'LICENSE',
        subtitle: randomDateString(start, end),
        uniqid: uuid(),
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
        type: nodeTypes.ITEM,
      },
      {
        title: 'Makefile',
        subtitle: randomDateString(start, end),
        uniqid: uuid(),
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
        type: nodeTypes.ITEM,
      },
      {
        title: 'package.json',
        subtitle: randomDateString(start, end),
        uniqid: uuid(),
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
        type: nodeTypes.ITEM,
      },
      {
        title: 'README.md',
        subtitle: randomDateString(start, end),
        uniqid: uuid(),
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
        type: nodeTypes.ITEM,
      },
      {
        title: 'webpack.config.js',
        subtitle: randomDateString(start, end),
        uniqid: uuid(),
        get id() {
          return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
        },
        type: nodeTypes.ITEM,
      },
    ],
  },
  {
    title: 'react-sortable-tree bottom',
    subtitle: randomDateString(start, end),
    uniqid: uuid(),
    get id() {
      return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
    },
    type: nodeTypes.FOLDER,
    children: [],
  },
  {
    title: 'QuillJS',
    subtitle: randomDateString(start, end),
    uniqid: uuid(),
    get id() {
      return `${this.type}${_ID_DELIMITER}${this.uniqid}`;
    },
    type: nodeTypes.ITEM,
  },
];

