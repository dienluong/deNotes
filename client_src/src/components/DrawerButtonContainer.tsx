import { connect } from 'react-redux';
import * as rootReducer from '../redux/reducers';
import { clickAction } from '../redux/actions/drawerButtonActions';
import DrawerButton from './widgets/DrawerButton';

// Types
import { AnyAction, Dispatch } from 'redux';
interface MapStatePropsT {
  visible: boolean,
}
interface MapDispatchPropsT {
  clickHandler: () => AnyAction;
}

function mapStateToProps(state: AppStateT): MapStatePropsT {
  return {
    visible: rootReducer.selectConnectionInfoLoggedIn(state) && !rootReducer.selectNotesTreeVisible(state),
  };
}

function mapDispatchToProps(dispatch: Dispatch<AnyAction>): MapDispatchPropsT {
  return {
    clickHandler() {
      return dispatch(clickAction());
    },
  }
}

// function mergeProps(stateProps: MapStatePropsT, dispatchProps: MapDispatchPropsT, ownProps: object) {
//   return Object.assign({}, dispatchProps, ownProps, { visible: stateProps.loggedIn && stateProps.buttonVisible });
// }

// Note: ownProps is automatically forwarded to the wrapped component
const DrawerButtonContainer = connect(mapStateToProps, mapDispatchToProps)(DrawerButton);
export default DrawerButtonContainer;

