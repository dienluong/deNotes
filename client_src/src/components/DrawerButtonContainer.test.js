import React from 'react';
import DrawerButtonContainer from './DrawerButtonContainer';
import DrawerButton from './widgets/DrawerButton';
import createMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { render, cleanup } from '@testing-library/react';
import 'jest-dom/extend-expect';
import initialState from '../redux/misc/initialState';
import { MutationObserver } from '../test-utils/MutationObserver';

// Configure Enzyme
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
let wrapper = {};

beforeAll(() => {
  // Required for rendering Quill editor
  global.MutationObserver = MutationObserver;
  document.getSelection = function() {
    return {
      getRangeAt: () => {},
    };
  };
});

afterEach(() => {
  if (typeof wrapper.unmount === 'function') {
    wrapper.unmount();
  }
  cleanup();
});

it('renders DrawerButton with "visible" prop set to false because loggedIn is false', () => {
  const mockedInitialState = {
    ...initialState,
    connectionInfo: {
      loggedIn: false,
    },
  };
  const dummyClickHandler = () => {};
  const dummyOwnProps = {
    size: 'medium',
    visible: true,
    clickHandler: dummyClickHandler,
    className: 'someClassName',
  };
  const expectedProps = {
    ...dummyOwnProps,
    visible: false,
  };
  const store = createMockStore([thunk])(mockedInitialState);
  const { container } = render(<Provider store={ store }><DrawerButtonContainer {...dummyOwnProps }/></Provider>);
  expect(container).toMatchSnapshot();

  wrapper = shallow(<Provider store={ store }><DrawerButtonContainer { ...dummyOwnProps }/></Provider>).dive({ context: { store } });
  expect(wrapper.find(DrawerButton).exists()).toBe(true);
  expect(wrapper.find(DrawerButton).props()).toMatchObject(expectedProps);
});

it('renders DrawerButton with "visible" prop taking the passed value because loggedIn is true', () => {
  const mockedInitialState = {
    ...initialState,
    connectionInfo: {
      loggedIn: true,
    },
  };
  const dummyClickHandler = () => {};
  // Case 1: passing visible = true
  let dummyOwnProps = {
    size: 'medium',
    visible: true,
    clickHandler: dummyClickHandler,
    className: 'someClassName',
  };
  let expectedProps = {
    ...dummyOwnProps,
    visible: true,
  };
  let store = createMockStore([thunk])(mockedInitialState);
  const { container } = render(<Provider store={ store }><DrawerButtonContainer {...dummyOwnProps }/></Provider>);
  expect(container).toMatchSnapshot();

  wrapper = shallow(<Provider store={ store }><DrawerButtonContainer { ...dummyOwnProps }/></Provider>).dive({ context: { store } });
  expect(wrapper.find(DrawerButton).exists()).toBe(true);
  expect(wrapper.find(DrawerButton).props()).toMatchObject(expectedProps);

  // Case 2: passing visible = false
  dummyOwnProps = {
    size: 'small',
    visible: false,
    clickHandler: dummyClickHandler,
    className: 'someClassName2',
  };
  expectedProps = {
    ...dummyOwnProps,
    visible: false,
  };

  store = createMockStore([thunk])(mockedInitialState);
  wrapper = shallow(<Provider store={ store }><DrawerButtonContainer { ...dummyOwnProps }/></Provider>).dive({ context: { store } });
  expect(wrapper.find(DrawerButton).exists()).toBe(true);
  expect(wrapper.find(DrawerButton).props()).toMatchObject(expectedProps);
});

