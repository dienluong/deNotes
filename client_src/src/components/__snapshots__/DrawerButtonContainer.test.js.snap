// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DrawerButton with "visible" prop set to false because loggedIn is false 1`] = `
<div>
  <div
    class="someClassName"
  />
</div>
`;

exports[`renders DrawerButton with "visible" prop taking the passed value because loggedIn is true 1`] = `
<div>
  <div
    class="someClassName"
  >
    <button
      aria-label="notes list drawer"
      class="MuiButtonBase-root-11 MuiFab-root-1 MuiFab-primary-3 MuiFab-sizeMedium-10"
      style="transform: scale(1); webkit-transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiFab-label-2"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root-14"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root-23"
      />
    </button>
  </div>
</div>
`;
